### PR TITLE
FNA returns LSE and per-query maximums!

### PR DESCRIPTION
**Pull request by AdityaKane2001**
FNA will now optionally return maximum per-query pre-softmax attention weights, along with LSE.

Changes:
* csrc: Added actual saving logic to kernel_forward.cuh and fna_forward.cuh. Changed function signature across NATTEN interface to accept an optional maximums tensor. Removed LSE as a template parameter.
* src: Added relevant arguments in Python interface and Function.forward()
* scripts: Changed autogen script to exclude LSE as a template parameter.
* tests: Added tests to confirm statistics (using bmm kernels sdpa).  

I have added the autogen'd files as well, let me know if you'd like me to create a PR without them to reduce clutter.

@alihassanijr please take a look when you get a moment. Let me know if I have missed anything, especially in the context of Python API surface.

### CODE REVIEW

File: csrc/include/natten/cuda/fna/fna_forward.cuh [hunk 1]
```diff
@@ -86,6 +87,7 @@ void fna_forward_generic(
 
   bool has_rpb = rpb_ptr != nullptr;
   bool compute_logsumexp = logsumexp_ptr != nullptr;
+  bool compute_maximums = maximums_ptr != nullptr;
```
Comment (alihassanijr) [inline csrc/include/natten/cuda/fna/fna_forward.cuh:90]: Unused?

File: csrc/include/natten/cuda/fna/fna_forward.cuh [hunk 2]
```diff
@@ -126,6 +128,9 @@ void fna_forward_generic(
     p.logsumexp_ptr = compute_logsumexp
         ? (typename Kernel::lse_scalar_t*)logsumexp_ptr
         : nullptr;
+    p.maximums_ptr = compute_maximums
+        ? (typename Kernel::max_scalar_t*)maximums_ptr
+        : nullptr;
```
Comment (alihassanijr) [inline csrc/include/natten/cuda/fna/fna_forward.cuh:133]:
```suggestion
    p.logsumexp_ptr = logsumexp_ptr != nullptr
        ? static_cast<typename Kernel::lse_scalar_t*>(logsumexp_ptr)
        : nullptr;
    p.maximums_ptr = compute_maximums
        ? static_cast<typename Kernel::max_scalar_t*>(maximums_ptr)
        : nullptr;
```

File: csrc/include/natten/cuda/fna/kernel_forward.h [hunk 3]
```diff
@@ -129,14 +128,14 @@ struct FusedNeighborhoodAttentionKernel {
...
   static constexpr bool kSupportsRPB = kSupportsRPB_;
-  static constexpr bool kStoresLSE = kStoresLSE_;
   static_assert(
       !kSupportsRPB || !kHasCausalDims,
       "Causal NA does not support RPB yet.");
 
   using scalar_t = scalar_t_;
   using accum_t = float;
   using lse_scalar_t = float;
+  using max_scalar_t = float;
```
Comment (alihassanijr) [inline csrc/include/natten/cuda/fna/kernel_forward.h:138]: Let's not introduce different element types for LSE and max. This was a mistake in the kernel to begin with.

It's assuming `lse_scalar_t` is the same as `accum_t`. Just replace both `lse_scalar_t` and `max_scalar_t` with `accum_t`, and remove these.

File: csrc/include/natten/cuda/fna/kernel_forward.h [hunk 4]
```diff
@@ -197,6 +197,7 @@ struct FusedNeighborhoodAttentionKernel {
     Dim key_tile_shape;
 
     Dim lse_strideM;
+    Dim maximums_strideM;
```
Comment (alihassanijr) [inline csrc/include/natten/cuda/fna/kernel_forward.h:200]: `lse_strideM` is equal to `maximums_strideM` with the conditions we force on these tensors. Remove and just replace with `stats_strideM` with a comment that it applies to both `lse` and `maximums`. Or just use `lse_strideM` for both.

File: csrc/include/natten/cuda/fna/kernel_forward.h [hunk 5]
```diff
@@ -288,24 +289,34 @@ struct FusedNeighborhoodAttentionKernel {
...
+      if (maximums_ptr != nullptr){
+          // Following same pattern as lse strides
+          auto maximums_stride_dilation = compute_stride(num_queries, num_heads);
+          maximums_strideM = maximums_stride_dilation * dilation;
+          maximums_ptr += batch_id * num_queries.prod32() * num_heads + 
+              ((first_query * maximums_strideM).sum() +
+               (dilation_idx * maximums_stride_dilation).sum()) +
               head_id;
-        }
       }
```
Comment (alihassanijr) [inline csrc/include/natten/cuda/fna/kernel_forward.h:316]: Can we simplify this a bit and compute the offset once separately, and add it to `logsumexp_ptr`, and then if `maximums_ptr != nullptr`, add exactly the same offset to the pointer. We shouldn't need to redo all of this.

File: csrc/include/natten/cuda/fna/kernel_forward.h [hunk 6]
```diff
@@ -1004,6 +1015,16 @@ struct FusedNeighborhoodAttentionKernel {
...
+    
+    if (p.maximums_ptr && thread_id() < kQueriesPerBlock){
+      constexpr float kLog2e = 1.4426950408889634074; // log_2(e) = M_LOG2E
+      auto query_idx =
+          map_index_to_coord((int32_t)thread_id(), problem_size_0_m);
+      auto query_offset = (query_idx * p.maximums_strideM).sum();
+      if (is_coord_within_upper_bound(query_idx, problem_size_0_m)) {
+        p.maximums_ptr[query_offset] = mi[thread_id()] / kLog2e; 
+      }
+    }
```
Comment (alihassanijr) [inline csrc/include/natten/cuda/fna/kernel_forward.h:1027]: Again a lot of code repetition here. Can we simplify and merge with the LSE block above?

File: csrc/include/natten/pytorch/helpers.h [hunk 7]
```diff
@@ -276,6 +276,47 @@ void CheckLogSumExp(const at::Tensor& output, const at::Tensor& logsumexp) {
...
+        "Invalid maximums shape at dim ",
+        i,
+        "; "
+        "expected ",
+        output.size(i),
+        ", got ",
+        maximums.size(i),
+        ".");
+  }
+}
```
Comment (alihassanijr) [inline csrc/include/natten/pytorch/helpers.h:318]: Please don't repeat this, and just rename the LSE one to something more generic, and use it for both LSE and maximums.

File: src/natten/functional.py [hunk 8]
```diff
@@ -1191,7 +1196,8 @@ def forward(
         ctx.tiling_config_backward = tiling_config_backward
         ctx.has_bias = bias is not None
 
-        return output
+        return output, logsumexp, maximums
```
Comment (alihassanijr) [inline src/natten/functional.py:1199]: Let's definitely not return either of `logsumexp` or `maximums` when `return_statistics` is `False`.

Comment (alihassanijr) [reply src/natten/functional.py:1199]: Actually nevermind, that doesn't make sense. I need to think more about how we're returning three things here.
